src/shim.c

```diff
@@ -151,100 +155,224 @@ err_exit(const char *format, ...)
+
+	if (header->payload_len) {
+		strncpy((char *)msg + PAYLOAD_OFFSET, (const char*)payload, header->payload_len);
+	}
 
-	return proxy_ctl_msg;
- }
+	return msg;
+}
 
 /*!
- * Send "hyper" payload to cc-proxy. This will be forwarded to hyperstart.
+ * Write header and payload to the proxy socket fd.
  *
- * \param fd File descriptor to send the message to(should be proxy ctl socket fd)
- * \param Hyperstart cmd id
- * \param json Json payload
+ * \param shim \ref cc_shim
+ * \param header \ref frame_header
+ * \param payload Payload to be sent
  */
-void
-send_proxy_hyper_message(int fd, const char *hyper_cmd, const char *json) {
-	char      *proxy_payload = NULL;
-	char      *proxy_command_id = "hyper";
-	char      *proxy_ctl_msg = NULL;
-	size_t     len = 0, offset = 0;
+bool
+write_message(struct cc_shim *shim, struct frame_header *header, 
+				const uint8_t *payload)
+{
+	size_t     offset = 0;
 	ssize_t    ret;
+	ssize_t    total_size;
 
-	/* cc-proxy has the following format for "hyper" payload:
-	 * {
-	 *    "id": "hyper",
-	 *    "data": {
-	 *        "hyperName": "ping",
-	 *        "data": "json payload",
-	 *    }
-	 * }
-	*/
-
-	if ( !json || fd < 0) {
-		return;
-	}
-
-
-	ret = asprintf(&proxy_payload,
-			"{\"id\":\"%s\",\"data\":{\"hyperName\":\"%s\",\"data\":%s}}",
-			proxy_command_id, hyper_cmd, json);
+	total_size = (ssize_t)((header->header_len * 32) + header->payload_len);
 
-	if (ret == -1) {
-		abort();
+	uint8_t *msg = serialize_message(header, payload, total_size);
+	if (! msg ) {
+		return false;
 	}
 
-	proxy_ctl_msg = get_proxy_ctl_msg(proxy_payload, &len);
-	free(proxy_payload);
-
-	while (offset < len) {
-		ret = write(fd, proxy_ctl_msg + offset, len-offset);
+	while (offset < total_size) {
+		ret = write(shim->proxy_sock_fd, msg + offset, (size_t)total_size-offset);
 		if (ret == EINTR) {
 			continue;
 		}
 		if (ret <= 0 ) {
-			free(proxy_ctl_msg);
+			free(msg);
 			shim_error("Error writing to proxy: %s\n", strerror(errno));
-			return;
+			return false;
 		}
 		offset += (size_t)ret;
 	}
-	free(proxy_ctl_msg);
+	free(msg);
+	return true;
+}
+
+/*!
+ * Send message to proxy.
+ *
+ * \param shim \ref cc_shim
+ * \param header \ref frame_header
+ * \param payload Payload to be sent
+ */
+bool
+send_proxy_message(struct cc_shim *shim, uint8_t type, uint8_t opcode,
+			const char *payload)
+{
+	struct frame_header header = { 0 };
+	bool ret;
+
+	if ( !shim || shim->proxy_sock_fd < 0) {
+		return false;
+	}
+
+	header.version = PROXY_VERSION;
+	header.header_len = PROXY_HEADER_SIZE;
+	header.type = type;
+	header.opcode = opcode;
+
+	header.payload_len = (uint32_t)strlen(payload);
+
+	ret = write_message(shim, &header, (uint8_t *)payload);
+	return ret;
+}
+
+/*!
+ * Send Connect command to proxy.
+ *
+ * \param shim \ref cc_shim
+ */
+bool
+send_proxy_connect_cmd(struct cc_shim *shim)
+{
+	char *payload = NULL;
+	bool ret;
+
+	if (! (shim && shim->token) || shim->proxy_sock_fd < 0) {
```

Comment:
I'd break this check up to make it clearer that the check on the `shim` pointer is fully handled before the checks on the members:

```
if (! shim) {
    return false;
}

if ((! shim->token) || (shim->proxy_sock_fd < 0)) {
    return false;
}
```